lib/bakeAmbientOcclusion.js

```diff
@@ -0,0 +1,586 @@
+
+        var color = defaultValue(diffuse, [1.0, 1.0, 1.0, 1.0]);
+        for (var i = 0; i < 4; i++) {
+            diffuse[i] *= 255;
+        }
+        var newJimpImage = allImages[newImageID].extras._pipeline.jimpImage;
+
+        var resolution = state.resolution;
+        newJimpImage.resize(resolution, resolution);
+        var hexColor = Jimp.rgbaToInt(color[0], color[1], color[2], color[3]);
+        for (var x = 0; x < resolution; x++) {
+            for (var y = 0; y < resolution; y++) {
+                newJimpImage.setPixelColor(x, y, hexColor);
+            }
+        }
+        primitive.material = newMaterialID;
+        return allImages[newImageID];
+    }
+
+    var textureID = diffuse;
+    var imageID = allTextures[textureID].source;
+
+    if (materialsSeen.hasOwnProperty(materialID)) {
+        // Check if the material is unique. If not, clone material, texture, and image
+        cloneAndSetup(
+            newMaterialID, newTextureID, newImageID,
+            materialID, textureID, imageID,
+            allMaterials, allTextures, allImages);
+        primitive.material = newMaterialID;
+    } else if(texturesSeen.hasOwnProperty(textureID)) {
+        // Check if the texture is unique. If not clone the texture and the image.
+        cloneAndSetup(
+            undefined, newTextureID, newImageID,
+            materialID, textureID, imageID,
+            allMaterials, allTextures, allImages);
+        values.diffuse = newTextureID;
+    } else if(imagesSeen.hasOwnProperty(imageID)) {
+        // Check if the image is unique. if not, clone the image.
+        var texture = allTextures[textureID];
+        cloneAndSetup(
+            undefined, undefined, newImageID,
+            materialID, textureID, imageID,
+            allMaterials, allTextures, allImages);
+        texture.source = newImageID;
+    } else {
+        // If nothing was cloned, mark this material, texture, and image as seen
+        materialsSeen[materialID] = true;
+        texturesSeen[textureID] = true;
+        imagesSeen[imageID] = true;
+        newImageID = imageID;
+    }
+    return allImages[newImageID];
+}
+
+////////// loading //////////
+
+function generateRaytracerScene(gltf, scene, options) {
+    // Set up data we need for sampling. generate "triangle soup" over the whole scene.
+    var accessors = gltf.accessors;
+
+    // read all accessors in one go to avoid repeated read-and-conversion
+    var bufferDataByAccessor = {};
+
+    for (var accessorID in accessors) {
+        if (accessors.hasOwnProperty(accessorID)) {
+            var accessor = accessors[accessorID];
+            bufferDataByAccessor[accessorID] = readAccessor(gltf, accessor);
+        }
+    }
+
+    var raytracerScene = {
+        bufferDataByAccessor: bufferDataByAccessor,
+        numberSamples: defaultValue(options.numberSamples, 16),
+        rayDepth: defaultValue(options.rayDepth, 1.0), // TODO: compute dynamic default ray depth?
```

Comment:
FYI - `TODOs` in `ao-master` are fine, but we would not merge them into master; we will either do them or move them to a roadmap.